Clean up Generate(), part 1

valadaptive opened the pull request:
I've modified the `Generate` function to return a promise within a promise. If you call `let p = await Generate(...)`, it'll wait until generation has *started*. If you then do `await p`, it'll wait until generation has *finished*.

This makes it much easier to tell when generation has finished, which we now use in generateGroupWrapper instead of the hack that was there before.

I've given this a basic [smoke test](https://en.wikipedia.org/wiki/Smoke_testing_(software)), but I'm not sure if this breaks any corner cases.

valadaptive committed:
Remove isGenerationAborted

Just check the AbortSignal.

valadaptive committed:
Always call resolve in Generate()

This lets us get rid of the janky hack in group-chats to tell when a
message is done generating.

valadaptive committed:
Have Generate() return a promise

Generate(), being async, now returns a promise-within-a-promise.
If called with `let p = await Generate(...)`, it'll wait for generation
to *start*. If you then `await p`, you'll wait for generation to
*finish*. This makes it much easier to tell exactly when generation's
done. generateGroupWrapper has been similarly modified.

Cohee1207 commented:
Needs a rebase/merge

valadaptive commented:
Rebased

Cohee1207 commented:
generateQuietPrompt is currently broken and doesn't return any text in result.

valadaptive committed:
Fix generateQuietPrompt

valadaptive committed:
Reject on data.error

valadaptive commented:
I forgot to return a value. Try it again?

Cohee1207 commented:
Okay, there's also a slightly different problem. If the first pass of Generate produces no text and it tries to regenerate, the promise resolves immediately but it loops generation in the background until it gets killed by the circuit breaker. I see two options: either remove the looping on empty text (since it doesn't even work for streaming) or make the promise reject only when the circuit breaker kicks in.

valadaptive committed:
Reject generation on circuit breaker error

valadaptive commented:
I've changed it to reject when the circuit breaker kicks in. I can probably modify quiet mode to work with the looping, but I need some clarification on the behavior first. It looks like the setTimeout waits for `generate_loop_counter * 1000` seconds, meaning that the first loop through it waits for 1000 seconds, then after that it waits for 2000 seconds, then 3000...

Is this "backoff" behavior intentional?

Cohee1207 commented:
That was initially intentional. But I think this backoff is not needed anymore since it was made to avoid spamming Poe with requests and Poe support was dropped in July.

valadaptive committed:
Remove looping backoff behavior

valadaptive committed:
Improve circuit breaker

We now track the loop counter as a parameter of Generate that we
decrement with every recursive call, rather than a global variable,
and it *should* now work with quiet prompt generation.

valadaptive commented:
The circuit breaker should work with quiet mode now; give it a try

Cohee1207 committed:
Typing indicator fixed

Cohee1207 committed:
Fix group wrapper not resolving to a valid text

Cohee1207 committed:
Return from Generate if calling circuit breaker

Cohee1207 commented:
Fixed some edge cases. Will test some more and merge

Cohee1207 committed:
Fix abort group generation

Cohee1207 commented:
Ok, think that's all. Can resolve minor bugs later (if any)
